## Aula 1

### criação de dataframes, .loc e .iloc

#### Importação

In [68]:
import pandas as pd
import numpy as np

#### O que são series?


é um array capaz de guardar quaisquer tipo de dados

In [2]:
# iniciando uma serie
series = pd.Series([1, 2, 3.0, 'Indiozinhos'])
series

0              1
1              2
2            3.0
3    Indiozinhos
dtype: object

In [3]:
# tipo de uma série
type(series)

pandas.core.series.Series

In [4]:
# valores em uma series
series.values

array([1, 2, 3.0, 'Indiozinhos'], dtype=object)

In [5]:
# indexes de uma series
series.index

RangeIndex(start=0, stop=4, step=1)

In [6]:
# operações com series
series[0:3].mean()

2.0

In [7]:
# tipos de itens na series
# list comprehension
[type(i) for i in series]

[int, int, float, str]

In [8]:
# mudando os indices de uma series
series.index = ['um', '2', 3, 'quatro']
series

um                  1
2                   2
3                 3.0
quatro    Indiozinhos
dtype: object

In [9]:
# acessar elementos pelo indice
series['quatro']

'Indiozinhos'

as colunas dos df são formadas por series

## lendo arquivos com pandas

In [10]:
#lendo o arquivo ./data/alunos.csv
tabela = pd.read_csv('./data/alunos.csv', sep=';')
tabela

RA             Nome  Frequencia  Prova_1  Prova_2  Prova_3  Prova_4
0  110201   Antonio Carlos          20      6.5      8.5      7.0        6
1  110212      Ana Beatriz          20      7.0      7.0      7.0        8
2  110218    Carlos Vernes          17      7.0      7.0      7.0        7
3  110307  Francisco Cunha          20      9.0      8.5      8.5       10
4  110275      Sandra Rosa          15      6.5      7.5      7.0        7
5  110281   Juliana Arruda          18      7.5      7.0      7.5        8
6  110301        Joao Galo          20      5.0      6.5      7.0        5
7  110263     José Valente          20     10.0     10.0     10.0       10
8  110271   Maria Ferreira          19      9.5      8.0      7.0       10
9  110236  Adriana Tavares          20      8.0      8.0      8.0        8

In [11]:
# if you want to enter columns names:
tabela2 = pd.read_csv('./data/alunos.csv', sep=';', names=['a', 'b', 'c', 'd', 'e', 'f', 'g'])
tabela2

a                b           c        d        e        f        g
0       RA             Nome  Frequencia  Prova_1  Prova_2  Prova_3  Prova_4
1   110201   Antonio Carlos          20      6.5      8.5        7        6
2   110212      Ana Beatriz          20        7        7        7        8
3   110218    Carlos Vernes          17        7        7        7        7
4   110307  Francisco Cunha          20        9      8.5      8.5       10
5   110275      Sandra Rosa          15      6.5      7.5        7        7
6   110281   Juliana Arruda          18      7.5        7      7.5        8
7   110301        Joao Galo          20        5      6.5        7        5
8   110263     José Valente          20       10       10       10       10
9   110271   Maria Ferreira          19      9.5        8        7       10
10  110236  Adriana Tavares          20        8        8        8        8

## Selecionando Colunas

In [12]:
# acessar coluna 'Nome' diretamente
tabela.Nome

0     Antonio Carlos
1        Ana Beatriz
2      Carlos Vernes
3    Francisco Cunha
4        Sandra Rosa
5     Juliana Arruda
6          Joao Galo
7       José Valente
8     Maria Ferreira
9    Adriana Tavares
Name: Nome, dtype: object

In [13]:
# acessar coluna 'Nome diretamente'
tabela['Nome']

0     Antonio Carlos
1        Ana Beatriz
2      Carlos Vernes
3    Francisco Cunha
4        Sandra Rosa
5     Juliana Arruda
6          Joao Galo
7       José Valente
8     Maria Ferreira
9    Adriana Tavares
Name: Nome, dtype: object

In [14]:
# acessar coluna 'Nome' diretamente
# quando vc passa com uma lista
tabela[['Nome']]

Nome
0   Antonio Carlos
1      Ana Beatriz
2    Carlos Vernes
3  Francisco Cunha
4      Sandra Rosa
5   Juliana Arruda
6        Joao Galo
7     José Valente
8   Maria Ferreira
9  Adriana Tavares

quando vc acessa com uma lista, o resultado é um df

In [15]:
# selecionando apenas algumas colunas do df
colunas = ['Nome', 'Prova_1']
tabela[colunas]

Nome  Prova_1
0   Antonio Carlos      6.5
1      Ana Beatriz      7.0
2    Carlos Vernes      7.0
3  Francisco Cunha      9.0
4      Sandra Rosa      6.5
5   Juliana Arruda      7.5
6        Joao Galo      5.0
7     José Valente     10.0
8   Maria Ferreira      9.5
9  Adriana Tavares      8.0

In [16]:
# acessando linhas e elementos

In [17]:
# selecionar todas as linhas das colunas ['Nome', 'Prova_1']
tabela.loc[:, ['Nome', 'Prova_1']]

Nome  Prova_1
0   Antonio Carlos      6.5
1      Ana Beatriz      7.0
2    Carlos Vernes      7.0
3  Francisco Cunha      9.0
4      Sandra Rosa      6.5
5   Juliana Arruda      7.5
6        Joao Galo      5.0
7     José Valente     10.0
8   Maria Ferreira      9.5
9  Adriana Tavares      8.0

In [18]:
# selecionar linha 3 (index 2), tabelas 'Nome' 'Prova'
tabela.loc[2, ['Nome', 'Prova_1']]

Nome       Carlos Vernes
Prova_1              7.0
Name: 2, dtype: object

In [19]:
# selecionar linhas 3 a 6 das colunas 'Nome' e 'Prova'
tabela.loc[2:5, ['Nome', 'Prova_1']]

Nome  Prova_1
2    Carlos Vernes      7.0
3  Francisco Cunha      9.0
4      Sandra Rosa      6.5
5   Juliana Arruda      7.5

dica: pense em linhas começcando da linha 0 (ZERO)

In [20]:
# selecionar da linha 0 a 4, das colunas 'Prova_1' a 'Prova_4'
tabela.loc[0:4, 'Prova_1':'Prova_4']

Prova_1  Prova_2  Prova_3  Prova_4
0      6.5      8.5      7.0        6
1      7.0      7.0      7.0        8
2      7.0      7.0      7.0        7
3      9.0      8.5      8.5       10
4      6.5      7.5      7.0        7

## Alterar Valores

In [21]:
# alterar o valor da linha 7, coluna nome para joaozinho
tabela.loc[7, 'Nome'] = 'Joãozinho'

### fazer uma copia do df

In [22]:
tabela2 = tabela.copy
tabela2

<bound method NDFrame.copy of        RA             Nome  Frequencia  Prova_1  Prova_2  Prova_3  Prova_4
0  110201   Antonio Carlos          20      6.5      8.5      7.0        6
1  110212      Ana Beatriz          20      7.0      7.0      7.0        8
2  110218    Carlos Vernes          17      7.0      7.0      7.0        7
3  110307  Francisco Cunha          20      9.0      8.5      8.5       10
4  110275      Sandra Rosa          15      6.5      7.5      7.0        7
5  110281   Juliana Arruda          18      7.5      7.0      7.5        8
6  110301        Joao Galo          20      5.0      6.5      7.0        5
7  110263        Joãozinho          20     10.0     10.0     10.0       10
8  110271   Maria Ferreira          19      9.5      8.0      7.0       10
9  110236  Adriana Tavares          20      8.0      8.0      8.0        8>

In [23]:
# mudar o index para uma coluna
tabela2 = tabela.set_index('Nome')
tabela2

RA  Frequencia  Prova_1  Prova_2  Prova_3  Prova_4
Nome                                                                   
Antonio Carlos   110201          20      6.5      8.5      7.0        6
Ana Beatriz      110212          20      7.0      7.0      7.0        8
Carlos Vernes    110218          17      7.0      7.0      7.0        7
Francisco Cunha  110307          20      9.0      8.5      8.5       10
Sandra Rosa      110275          15      6.5      7.5      7.0        7
Juliana Arruda   110281          18      7.5      7.0      7.5        8
Joao Galo        110301          20      5.0      6.5      7.0        5
Joãozinho        110263          20     10.0     10.0     10.0       10
Maria Ferreira   110271          19      9.5      8.0      7.0       10
Adriana Tavares  110236          20      8.0      8.0      8.0        8

o metodo .loc ó funciona com o index, nesse caso, os index são os nomes das tabelas
neste caso podemos usar o metodo .iloc

In [24]:
tabela2.iloc[3: 5, [0, 1]]

RA  Frequencia
Nome                               
Francisco Cunha  110307          20
Sandra Rosa      110275          15

iloc seleciona atraves das linhas e colunas

In [25]:
# resetando o index da tabela2
tabela2.reset_index(inplace=True)
tabela2

Nome      RA  Frequencia  Prova_1  Prova_2  Prova_3  Prova_4
0   Antonio Carlos  110201          20      6.5      8.5      7.0        6
1      Ana Beatriz  110212          20      7.0      7.0      7.0        8
2    Carlos Vernes  110218          17      7.0      7.0      7.0        7
3  Francisco Cunha  110307          20      9.0      8.5      8.5       10
4      Sandra Rosa  110275          15      6.5      7.5      7.0        7
5   Juliana Arruda  110281          18      7.5      7.0      7.5        8
6        Joao Galo  110301          20      5.0      6.5      7.0        5
7        Joãozinho  110263          20     10.0     10.0     10.0       10
8   Maria Ferreira  110271          19      9.5      8.0      7.0       10
9  Adriana Tavares  110236          20      8.0      8.0      8.0        8

.iloc traz a linha
.loc traz o indice

## Criar novas colunas

In [26]:
tabela2['Mean'] = tabela2['Prova_1']/2
tabela2

Nome      RA  Frequencia  Prova_1  Prova_2  Prova_3  Prova_4  \
0   Antonio Carlos  110201          20      6.5      8.5      7.0        6   
1      Ana Beatriz  110212          20      7.0      7.0      7.0        8   
2    Carlos Vernes  110218          17      7.0      7.0      7.0        7   
3  Francisco Cunha  110307          20      9.0      8.5      8.5       10   
4      Sandra Rosa  110275          15      6.5      7.5      7.0        7   
5   Juliana Arruda  110281          18      7.5      7.0      7.5        8   
6        Joao Galo  110301          20      5.0      6.5      7.0        5   
7        Joãozinho  110263          20     10.0     10.0     10.0       10   
8   Maria Ferreira  110271          19      9.5      8.0      7.0       10   
9  Adriana Tavares  110236          20      8.0      8.0      8.0        8   

   Mean  
0  3.25  
1  3.50  
2  3.50  
3  4.50  
4  3.25  
5  3.75  
6  2.50  
7  5.00  
8  4.75  
9  4.00

In [27]:
# criar coluna np.nan
tabela2['nans'] = np.nan
tabela2

Nome      RA  Frequencia  Prova_1  Prova_2  Prova_3  Prova_4  \
0   Antonio Carlos  110201          20      6.5      8.5      7.0        6   
1      Ana Beatriz  110212          20      7.0      7.0      7.0        8   
2    Carlos Vernes  110218          17      7.0      7.0      7.0        7   
3  Francisco Cunha  110307          20      9.0      8.5      8.5       10   
4      Sandra Rosa  110275          15      6.5      7.5      7.0        7   
5   Juliana Arruda  110281          18      7.5      7.0      7.5        8   
6        Joao Galo  110301          20      5.0      6.5      7.0        5   
7        Joãozinho  110263          20     10.0     10.0     10.0       10   
8   Maria Ferreira  110271          19      9.5      8.0      7.0       10   
9  Adriana Tavares  110236          20      8.0      8.0      8.0        8   

   Mean  nans  
0  3.25   NaN  
1  3.50   NaN  
2  3.50   NaN  
3  4.50   NaN  
4  3.25   NaN  
5  3.75   NaN  
6  2.50   NaN  
7  5.00   NaN  
8  4.75   NaN  
9  4.00   NaN

tambem é possivel criar uma nova linha passando todos os das colunas

In [28]:
tabela2.loc[10, :] = ['tulio',2, 3, 4, 5, 6, 7, 'Prova_1', np.nan]
tabela2

Nome        RA  Frequencia  Prova_1  Prova_2  Prova_3  Prova_4  \
0    Antonio Carlos  110201.0        20.0      6.5      8.5      7.0      6.0   
1       Ana Beatriz  110212.0        20.0      7.0      7.0      7.0      8.0   
2     Carlos Vernes  110218.0        17.0      7.0      7.0      7.0      7.0   
3   Francisco Cunha  110307.0        20.0      9.0      8.5      8.5     10.0   
4       Sandra Rosa  110275.0        15.0      6.5      7.5      7.0      7.0   
5    Juliana Arruda  110281.0        18.0      7.5      7.0      7.5      8.0   
6         Joao Galo  110301.0        20.0      5.0      6.5      7.0      5.0   
7         Joãozinho  110263.0        20.0     10.0     10.0     10.0     10.0   
8    Maria Ferreira  110271.0        19.0      9.5      8.0      7.0     10.0   
9   Adriana Tavares  110236.0        20.0      8.0      8.0      8.0      8.0   
10            tulio       2.0         3.0      4.0      5.0      6.0      7.0   

       Mean  nans  
0      3.25   NaN  
1       3.5   NaN  
2       3.5   NaN  
3       4.5   NaN  
4      3.25   NaN  
5      3.75   NaN  
6       2.5   NaN  
7       5.0   NaN  
8      4.75   NaN  
9       4.0   NaN  
10  Prova_1   NaN

## Métodos prontos

importante prestar ateção nos axis

In [29]:
tabela2.loc[:,'Prova_1':'Prova_4'].mean(axis=1)

0      7.000
1      7.250
2      7.000
3      9.000
4      7.000
5      7.500
6      5.875
7     10.000
8      8.625
9      8.000
10     5.500
dtype: float64

In [30]:
tabela2['Mean']= tabela2.loc[:,'Prova_1':'Prova_4'].mean(axis=1)
tabela2

Nome        RA  Frequencia  Prova_1  Prova_2  Prova_3  Prova_4  \
0    Antonio Carlos  110201.0        20.0      6.5      8.5      7.0      6.0   
1       Ana Beatriz  110212.0        20.0      7.0      7.0      7.0      8.0   
2     Carlos Vernes  110218.0        17.0      7.0      7.0      7.0      7.0   
3   Francisco Cunha  110307.0        20.0      9.0      8.5      8.5     10.0   
4       Sandra Rosa  110275.0        15.0      6.5      7.5      7.0      7.0   
5    Juliana Arruda  110281.0        18.0      7.5      7.0      7.5      8.0   
6         Joao Galo  110301.0        20.0      5.0      6.5      7.0      5.0   
7         Joãozinho  110263.0        20.0     10.0     10.0     10.0     10.0   
8    Maria Ferreira  110271.0        19.0      9.5      8.0      7.0     10.0   
9   Adriana Tavares  110236.0        20.0      8.0      8.0      8.0      8.0   
10            tulio       2.0         3.0      4.0      5.0      6.0      7.0   

      Mean  nans  
0    7.000   NaN  
1    7.250   NaN  
2    7.000   NaN  
3    9.000   NaN  
4    7.000   NaN  
5    7.500   NaN  
6    5.875   NaN  
7   10.000   NaN  
8    8.625   NaN  
9    8.000   NaN  
10   5.500   NaN

## Filtros

existem 3 principais metodos para fazer filtros.
dica do professor: seja muito bom em 1 método

In [31]:
# método 1
# retornar o sub-dataframe que contém valores maiores que 6 na coluna média
tabela2[tabela2['Mean']>7]

Nome        RA  Frequencia  Prova_1  Prova_2  Prova_3  Prova_4  \
1      Ana Beatriz  110212.0        20.0      7.0      7.0      7.0      8.0   
3  Francisco Cunha  110307.0        20.0      9.0      8.5      8.5     10.0   
5   Juliana Arruda  110281.0        18.0      7.5      7.0      7.5      8.0   
7        Joãozinho  110263.0        20.0     10.0     10.0     10.0     10.0   
8   Maria Ferreira  110271.0        19.0      9.5      8.0      7.0     10.0   
9  Adriana Tavares  110236.0        20.0      8.0      8.0      8.0      8.0   

     Mean  nans  
1   7.250   NaN  
3   9.000   NaN  
5   7.500   NaN  
7  10.000   NaN  
8   8.625   NaN  
9   8.000   NaN

## filtros compostos
Obs.: temos os seguintes operadores lógicos:

- &     - corresponde ao "and"
- |     - corresponde ao "or"
- ~     - corresponde ao "not"

In [32]:
# filtrar tabela2 para media menor que 7 e frequencia >= 20
tabela2[(tabela2['Mean']<7) & (tabela2['Frequencia']>=20)]

Nome        RA  Frequencia  Prova_1  Prova_2  Prova_3  Prova_4   Mean  \
6  Joao Galo  110301.0        20.0      5.0      6.5      7.0      5.0  5.875   

   nans  
6   NaN

In [33]:
# pegando só a coluna 'Mean' dos alunos que tiveram nota > 7 na prova 1 e frequencia maior ou igual a 20
tabela2[(tabela2['Prova_1']>7) & (tabela2['Frequencia']>=20)]['Mean']

3     9.0
7    10.0
9     8.0
Name: Mean, dtype: float64

In [34]:
# para mim é mais legivel selecionar primeiro as tabelas e depois os filtros
tabela2['Mean'][tabela2['Prova_1']<5]

10    5.5
Name: Mean, dtype: float64

In [35]:
# calculando a media de todos os alunos que tiveram nota maior que 7
# na prova 1 e frequencia maior ou igual a 20
tabela2[(tabela2['Prova_1']>7) & (tabela2['Frequencia']>=20)]['Mean'].mean()

9.0

In [36]:
# criando novo df direto de um filtro
prova_4 = tabela2[(tabela2['Prova_4']>=7) & (tabela2['Prova_3']>=6)]
prova_4

Nome        RA  Frequencia  Prova_1  Prova_2  Prova_3  Prova_4  \
1       Ana Beatriz  110212.0        20.0      7.0      7.0      7.0      8.0   
2     Carlos Vernes  110218.0        17.0      7.0      7.0      7.0      7.0   
3   Francisco Cunha  110307.0        20.0      9.0      8.5      8.5     10.0   
4       Sandra Rosa  110275.0        15.0      6.5      7.5      7.0      7.0   
5    Juliana Arruda  110281.0        18.0      7.5      7.0      7.5      8.0   
7         Joãozinho  110263.0        20.0     10.0     10.0     10.0     10.0   
8    Maria Ferreira  110271.0        19.0      9.5      8.0      7.0     10.0   
9   Adriana Tavares  110236.0        20.0      8.0      8.0      8.0      8.0   
10            tulio       2.0         3.0      4.0      5.0      6.0      7.0   

      Mean  nans  
1    7.250   NaN  
2    7.000   NaN  
3    9.000   NaN  
4    7.000   NaN  
5    7.500   NaN  
7   10.000   NaN  
8    8.625   NaN  
9    8.000   NaN  
10   5.500   NaN

Outras formas de filtrar

.query

## métodos
## 8. Métodos:

O Pandas possui diversos métodos que podem ser utilizados nessa estrutura.
Abaixo estão alguns métodos que essa estrutura de dados possui e facilitam alguns cálculos e análises [link](https://pandas.pydata.org/pandas-docs/version/0.23/api.html#api-panel-stats):
 

| Método      | Descrição     |
| ----------- | -----------   |
| sum         | soma          |
| mean        | média         |
| std         | desvio padrão |
| mode        | moda          |
| max         | valor máximo  |
| min         | valor mínimo  |
| idxmax      | primeiro índice com valor máximo |
| idxmin      | primeiro índice com valor mínimo |
| cumsum | soma cumulativa |
| cumprod | produto cumulativo |
| abs    | valor absoluto |



# Aula 2

## Leitura de dados
[Documentação](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html)

| Data | Reader | Writer|
|------|--------|-------|
| CSV	| read_csv | to_csv |
| JSON	| read_json | to_json |
| HTML	| read_html | to_html |
| XML	| read_xml | to_xml |
| Local clipboard |	read_clipboard | to_clipboard
| MS Excel	|read_excel	| to_excel |
| OpenDocument|	read_excel	| |
| HDF5 |	read_hdf |	to_hdf |
| Parquet | read_parquet | to_parquet |
| SAS	|read_sas	| |
| SPSS	|read_spss	| |
| Python Pickle |	read_pickle | to_pickle| 
| SQL	|read_sql|	to_sql |
| Google BigQuery|	read_gbq | to_gbq |


In [37]:
import pandas as pd
import numpy as np
import os

In [38]:
# encontrar path
os.getcwd()

'G:\\Meu Drive\\001_letscode\\DS_Portfolio\\meu_portfolio_2\\Lets_Code_DS_Degree_Alunos\\Módulo 3 - Pandas'

In [39]:
os.listdir()

['Aula_1_-_Criação_df_loc_iloc.ipynb',
 'Aula_2_-_Leitura_Filtragem_Dados_Métodos_ùteis.ipynb',
 'Aula_3_-_Limpeza_Dados_Dummies.ipynb',
 'concat_exemplo_colunas.png',
 'concat_exemplo_linhas.png',
 'groupby.png',
 'join_exemplo.png',
 'melt.png',
 'notas.xlsx',
 'pandas-data-structure.svg',
 'pandas_2_markdown.jpeg',
 'saida_iris.csv',
 'tabela_processada.xlsx',
 'variaveis_categoricas.jpeg',
 'Meu_resumo.ipynb',
 'Aula_4_-_Manipulação_Groupby_Merge.ipynb',
 'join_exemplo2.png',
 'data',
 '.ipynb_checkpoints']

In [40]:
# ler df do arquivo ./data/titanic.csv
df3 = pd.read_csv('./data/titanic.csv', sep=',')
df3.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

## Atributos

In [41]:
# .shape
df3.shape[1] 
# [1] para colunas
# [0] para linhas

12

In [42]:
df3.shape[0]

891

In [43]:
df3.shape

(891, 12)

In [44]:
# .size
# retorna a quantidade de elementos do df
df3.size

10692

In [45]:
# .columns
# retorna o nome das colunas e pode ser usado para renomear colunas
df3.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [46]:
df3.columns = ['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age',
               'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']
df3.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [47]:
# .index
# retorna o index do df
df3.index

RangeIndex(start=0, stop=891, step=1)

In [48]:
# .dtypes
df3.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

## Métodos 2
.head()
.tail()

In [49]:
df3.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

## .describe
descreve estatísticas básicas sobre todas as colunas

## df.info()
fornece a quantidade de valores nao nulos, o tipo de cada coluna e uso
de memória

## .astype()
converte o tipo da coluna

In [50]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [51]:
df3[['Pclass', 'Survived']] = df3[['Pclass', 'Survived']].astype('int8', str)
df3.dtypes
# coonverte para um tipo só

PassengerId      int64
Survived          int8
Pclass            int8
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

## .value_counts()
traz a contagem de elementos para cada valor distinto da coluna em que 
está sendo aplicado

In [52]:
# valores para Pclass
df3.Pclass.value_counts()

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [53]:
# pode usar uma lista de colunas
df3[['Pclass', 'Embarked']].value_counts()

Pclass  Embarked
3       S           353
2       S           164
1       S           127
        C            85
3       Q            72
        C            66
2       C            17
        Q             3
1       Q             2
dtype: int64

## .unique()
Retorna quem sãoos valores únicos da sua coluna

In [54]:
df3.Embarked.unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [55]:
# para mais de uma coluna
df3.Pclass.unique()

array([3, 1, 2], dtype=int8)

In [56]:
## .nunique
df3.Pclass.nunique()

3

## `.rename`
é possível renomear tanto colunas como index trocado pelo metodo de renomear index

In [57]:
# criando uma copia do df
df4 = df3.copy()
troca_nomes = {'Age': 'Idade', 'Name':'Nome'}
df4.rename(troca_nomes, axis=1, inplace=True)
df4.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Nome     Sex  Idade  SibSp  \
0                            Braund, Mr. Owen Harris    male   22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   38.0      1   
2                             Heikkinen, Miss. Laina  female   26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   35.0      1   
4                           Allen, Mr. William Henry    male   35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

## `.drop()`
permite deletar linhas ou colunas

In [58]:
df4.drop(['Sex', 'Idade'], axis=1)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Nome  SibSp  Parch  \
0                              Braund, Mr. Owen Harris      1      0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...      1      0   
2                               Heikkinen, Miss. Laina      0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)      1      0   
4                             Allen, Mr. William Henry      0      0   
..                                                 ...    ...    ...   
886                              Montvila, Rev. Juozas      0      0   
887                       Graham, Miss. Margaret Edith      0      0   
888           Johnston, Miss. Catherine Helen "Carrie"      1      2   
889                              Behr, Mr. Karl Howell      0      0   
890                                Dooley, Mr. Patrick      0      0   

               Ticket     Fare Cabin Embarked  
0           A/5 21171   7.2500   NaN        S  
1            PC 17599  71.2833   C85        C  
2    STON/O2. 3101282   7.9250   NaN        S  
3              113803  53.1000  C123        S  
4              373450   8.0500   NaN        S  
..                ...      ...   ...      ...  
886            211536  13.0000   NaN        S  
887            112053  30.0000   B42        S  
888        W./C. 6607  23.4500   NaN        S  
889            111369  30.0000  C148        C  
890            370376   7.7500   NaN        Q  

[891 rows x 10 columns]

## `.sort_values()`
para ordem reversa use `ascending=True`
voce pode usar mais de uma coluna para sort, inclusive usar ordem distintas para cada colune

In [59]:
df3.sort_values(['Pclass', 'Fare'], ascending=[True, False])

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1              Lesurer, Mr. Gustave J   
27            28         0       1      Fortune, Mr. Charles Alexander   
88            89         1       1          Fortune, Miss. Mabel Helen   
..           ...       ...     ...                                 ...   
378          379         0       3                 Betros, Mr. Tannous   
179          180         0       3                 Leonard, Mr. Lionel   
271          272         1       3        Tornquist, Mr. William Henry   
302          303         0       3     Johnson, Mr. William Cahoone Jr   
597          598         0       3                 Johnson, Mr. Alfred   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  
258  female  35.0      0      0  PC 17755  512.3292          NaN        C  
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C  
737    male  35.0      0      0  PC 17755  512.3292         B101        C  
27     male  19.0      3      2     19950  263.0000  C23 C25 C27        S  
88   female  23.0      3      2     19950  263.0000  C23 C25 C27        S  
..      ...   ...    ...    ...       ...       ...          ...      ...  
378    male  20.0      0      0      2648    4.0125          NaN        C  
179    male  36.0      0      0      LINE    0.0000          NaN        S  
271    male  25.0      0      0      LINE    0.0000          NaN        S  
302    male  19.0      0      0      LINE    0.0000          NaN        S  
597    male  49.0      0      0      LINE    0.0000          NaN        S  

[891 rows x 12 columns]

## `.memory_usage()`
retorna a quantifade de memoria utilizada por cada coluna em bytes

## `.set_index()`
escolhe uma coluna para ser o novo index
## `.reset_index()`
reseta o index para melhor

# Aula 3

- limpeza de dados
- trasformação de dados

In [60]:
import pandas as pd
import numpy as np

In [61]:
df = pd.read_csv('./data/titanic.csv')

In [62]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

## limpeza de dados faltantes

`isna()` and `isnull()`  
retornam um booleano se existe nulo naquela posição  
tanto `.isnull` como `.isna()` retornam um **df** boleano

In [63]:
df.isna()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Cabin  Embarked  
0    False   True     False  
1    False  False     False  
2    False   True     False  
3    False  False     False  
4    False   True     False  
..     ...    ...       ...  
886  False   True     False  
887  False  False     False  
888  False   True     False  
889  False  False     False  
890  False   True     False  

[891 rows x 12 columns]

In [64]:
# contar a quantidade de nulos
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## `notnull()`

In [65]:
df.notnull().sum()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

## `.dropna()`  
elimina nulos em todo o df ou em colunas especificas quando utilizado o argumento `subset`. aceita o parametro `inplace`

In [66]:
df.shape

(891, 12)

In [67]:
df.dropna(subset=['Age', 'Embarked']).shape

(712, 12)

In [68]:
df.dropna(subset=['Age', 'Embarked']).isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          529
Embarked         0
dtype: int64

## `.fillna()`  
preenche os valores nulos com valores especificados. Parâmtros úteis:  
- method {'bfill', 'ffill', none}, default None  
ffill preenche com o ultimo valor válido  
bfill preenche cm o próximo valor válido  
- axis {0 para index, 1 para colunas}  
eixo no quall se quer preencher os nulos  
- inplace

In [69]:
fill = pd.DataFrame({'original': [np.nan, 1, 2, np.nan, 4, 5, np.nan, np.nan, np.nan]})
fill

original
0       NaN
1       1.0
2       2.0
3       NaN
4       4.0
5       5.0
6       NaN
7       NaN
8       NaN

In [70]:
fill.original.mean()

3.0

In [71]:
#option 1
fill['cte'] = fill.original.fillna(999)
fill

original    cte
0       NaN  999.0
1       1.0    1.0
2       2.0    2.0
3       NaN  999.0
4       4.0    4.0
5       5.0    5.0
6       NaN  999.0
7       NaN  999.0
8       NaN  999.0

In [72]:
fill['media'] = fill.original.fillna(fill.original.mean())
fill

original    cte  media
0       NaN  999.0    3.0
1       1.0    1.0    1.0
2       2.0    2.0    2.0
3       NaN  999.0    3.0
4       4.0    4.0    4.0
5       5.0    5.0    5.0
6       NaN  999.0    3.0
7       NaN  999.0    3.0
8       NaN  999.0    3.0

In [73]:
fill['ffill'] = fill.original.fillna(method = 'ffill')
fill

original    cte  media  ffill
0       NaN  999.0    3.0    NaN
1       1.0    1.0    1.0    1.0
2       2.0    2.0    2.0    2.0
3       NaN  999.0    3.0    2.0
4       4.0    4.0    4.0    4.0
5       5.0    5.0    5.0    5.0
6       NaN  999.0    3.0    5.0
7       NaN  999.0    3.0    5.0
8       NaN  999.0    3.0    5.0

In [74]:
fill['ffilll_limit'] = fill.original.fillna(method='ffill', limit=2)
fill

original    cte  media  ffill  ffilll_limit
0       NaN  999.0    3.0    NaN           NaN
1       1.0    1.0    1.0    1.0           1.0
2       2.0    2.0    2.0    2.0           2.0
3       NaN  999.0    3.0    2.0           2.0
4       4.0    4.0    4.0    4.0           4.0
5       5.0    5.0    5.0    5.0           5.0
6       NaN  999.0    3.0    5.0           5.0
7       NaN  999.0    3.0    5.0           5.0
8       NaN  999.0    3.0    5.0           NaN

In [75]:
fill['bfill'] = fill.original.fillna(method='bfill')
fill

original    cte  media  ffill  ffilll_limit  bfill
0       NaN  999.0    3.0    NaN           NaN    1.0
1       1.0    1.0    1.0    1.0           1.0    1.0
2       2.0    2.0    2.0    2.0           2.0    2.0
3       NaN  999.0    3.0    2.0           2.0    4.0
4       4.0    4.0    4.0    4.0           4.0    4.0
5       5.0    5.0    5.0    5.0           5.0    5.0
6       NaN  999.0    3.0    5.0           5.0    NaN
7       NaN  999.0    3.0    5.0           5.0    NaN
8       NaN  999.0    3.0    5.0           NaN    NaN

## `interpolate()`  
faz a interpolação dos missing de acordo como o metodo escolhido.  

In [76]:
s = pd.DataFrame([0, 1, np.nan, 3, 4, np.nan, 400], columns=['original'])
s['linear'] = s.original.interpolate()
s['segundo_grau'] = s.original.interpolate(method='cubic', order=3)
s

original  linear  segundo_grau
0       0.0     0.0      0.000000
1       1.0     1.0      1.000000
2       NaN     2.0      5.367521
3       3.0     3.0      3.000000
4       4.0     4.0      4.000000
5       NaN   202.0     99.290598
6     400.0   400.0    400.000000

## `.replace()`  
substitui elementos dentro de um df

In [77]:
# para apenas uma troca
df.replace({'male':'h'}, inplace=True)

In [78]:
df.Sex.replace({'h':'H', 'female':'M'})

0      H
1      M
2      M
3      M
4      H
      ..
886    H
887    M
888    M
889    H
890    H
Name: Sex, Length: 891, dtype: object

In [79]:
df.Sex.replace({'h':'male'}, inplace=True)

In [80]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [81]:
# para mais de uma troca
df.replace({'Sex':{'male':'H', 'female':'F'}, 'Name':{df.Name[0]:'Fera'}})

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name Sex   Age  SibSp  \
0                                                 Fera   H  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...   F  38.0      1   
2                               Heikkinen, Miss. Laina   F  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)   F  35.0      1   
4                             Allen, Mr. William Henry   H  35.0      0   
..                                                 ...  ..   ...    ...   
886                              Montvila, Rev. Juozas   H  27.0      0   
887                       Graham, Miss. Margaret Edith   F  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"   F   NaN      1   
889                              Behr, Mr. Karl Howell   H  26.0      0   
890                                Dooley, Mr. Patrick   H  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [82]:
df.replace({'Embarked':
            {'C':'Cherbourg', 'Q':'Queenstown', 'S':'Southampton'}
           })['Embarked'].fillna(df.Embarked.mode(), inplace=True)

In [83]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## limpeza de dados: dados duplicados

## `.duplicate()` and `.drop_duplicated()`

o `.duplicated()` retorna series indicando se determinada linha possui duplicados ou não.  
e `.drop_duplicated()` elimina as linhas duplicadas  
- `subset` seleciona colunas para serem utilizadas na comparação de linhas duplicadas  
- `keep` {'first', 'last', 'False'} determina qual duplicado manter  
- `inplace`

In [85]:
df.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Length: 891, dtype: bool

In [86]:
df.drop_duplicates(subset=['Age', 'Embarked'])

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
5              6         0       3   
..           ...       ...     ...   
829          830         1       1   
843          844         0       3   
851          852         0       3   
852          853         0       3   
885          886         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
5                                     Moran, Mr. James    male   NaN      0   
..                                                 ...     ...   ...    ...   
829          Stone, Mrs. George Nelson (Martha Evelyn)  female  62.0      0   
843                         Lemberopolous, Mr. Peter L    male  34.5      0   
851                                Svensson, Mr. Johan    male  74.0      0   
852                            Boulos, Miss. Nourelain  female   9.0      1   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
5        0            330877   8.4583   NaN        Q  
..     ...               ...      ...   ...      ...  
829      0            113572  80.0000   B28      NaN  
843      0              2683   6.4375   NaN        C  
851      0            347060   7.7750   NaN        S  
852      1              2678  15.2458   NaN        C  
885      5            382652  29.1250   NaN        Q  

[167 rows x 12 columns]

In [87]:
# exercicio
# Elimine os duplicados considerando as colunas 'brand'e 'style', mantendo o ultimo valor

df_exerc = pd.DataFrame({
    'brand': ['Yum Yum', 'Yum Yum', 'Yum Yum', 'Indomie', 'Indomie', 'Indomie'],
    'style': ['cup', 'cup', 'cup', 'cup', 'pack', 'pack'],
    'rating': [4, 4, np.nan, 3.5, 15, 5]
})

In [88]:
df_exerc.head()

brand style  rating
0  Yum Yum   cup     4.0
1  Yum Yum   cup     4.0
2  Yum Yum   cup     NaN
3  Indomie   cup     3.5
4  Indomie  pack    15.0

In [94]:
# checado se tem duplicados
df_exerc.duplicated()

0    False
1     True
2    False
3    False
4    False
5    False
dtype: bool

In [95]:
# drop duplicates with last
df_exerc.drop_duplicates(keep='last')

brand style  rating
1  Yum Yum   cup     4.0
2  Yum Yum   cup     NaN
3  Indomie   cup     3.5
4  Indomie  pack    15.0
5  Indomie  pack     5.0

## transformação de dados  
## `get_dummies()`

### variaveis categoricas
variáveis categoricas são aquelas que representam grupos ou classes dentro dos nossos dados  
- ordinais: possuem uma ordem que tem um sentido  
- nominais: não possuem uma ordem lógica  
  
<img src="variaveis_categoricas.jpeg" style="width: 500px">
<a href="https://pandas.pydata.org/docs/reference/api/pandas.get_dummies.html">Parâmetros:</a> <br>
`pandas.get_dummies(data, prefix=None, prefx_sep='_', dummy_na=False, columns=None, sparse=False, drop_first=False, dtype=None`

In [96]:
pd.get_dummies(df, columns=['Sex', 'Embarked'], drop_first=True)

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name   Age  SibSp  Parch  \
0                              Braund, Mr. Owen Harris  22.0      1      0   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0      1      0   
2                               Heikkinen, Miss. Laina  26.0      0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0      1      0   
4                             Allen, Mr. William Henry  35.0      0      0   
..                                                 ...   ...    ...    ...   
886                              Montvila, Rev. Juozas  27.0      0      0   
887                       Graham, Miss. Margaret Edith  19.0      0      0   
888           Johnston, Miss. Catherine Helen "Carrie"   NaN      1      2   
889                              Behr, Mr. Karl Howell  26.0      0      0   
890                                Dooley, Mr. Patrick  32.0      0      0   

               Ticket     Fare Cabin  Sex_male  Embarked_Q  Embarked_S  
0           A/5 21171   7.2500   NaN         1           0           1  
1            PC 17599  71.2833   C85         0           0           0  
2    STON/O2. 3101282   7.9250   NaN         0           0           1  
3              113803  53.1000  C123         0           0           1  
4              373450   8.0500   NaN         1           0           1  
..                ...      ...   ...       ...         ...         ...  
886            211536  13.0000   NaN         1           0           1  
887            112053  30.0000   B42         0           0           1  
888        W./C. 6607  23.4500   NaN         0           0           1  
889            111369  30.0000  C148         1           0           0  
890            370376   7.7500   NaN         1           1           0  

[891 rows x 13 columns]

## material de aprofundamento

`.cut()`  
ordena os dados, separa em bins e computa qual grupo cada linha do df pertence. o `.cut()` escolherá os bins para serem espaçados uniformemente de acordo com os proprios valores e não coma  frequência desses valores.  
ele é muito utilizado para transformar variáveis contúnuas em categóricas. por exemplo, podemos converter o valor numérico da idade em grupos de criança, jovem, adulto e idoso.  
Parâmetros  
`pandas.cut(x, bins, right=True, labels= None, retbins=False, precision=3, include_lowest=False, duplicates='raise', ordered=True)`  
Ao informar a quantidade de grupos o pd.cut() escolhe os bins com o mesmo tamanho de janela:

In [99]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [107]:
df['cut_bins'] = pd.cut(df.Age, 4, precision=0, ordered=True)
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked      cut_bins  
0      0         A/5 21171   7.2500   NaN        S  (20.0, 40.0]  
1      0          PC 17599  71.2833   C85        C  (20.0, 40.0]  
2      0  STON/O2. 3101282   7.9250   NaN        S  (20.0, 40.0]  
3      0            113803  53.1000  C123        S  (20.0, 40.0]  
4      0            373450   8.0500   NaN        S  (20.0, 40.0]

In [108]:
df.cut_bins.value_counts()

(20.0, 40.0]    385
(0.0, 20.0]     179
(40.0, 60.0]    128
(60.0, 80.0]     22
Name: cut_bins, dtype: int64

Também podemos parrar uma lista com os valores de início e fim dos bins:

In [109]:
pd.cut(df.Age, [0, 20, 60, 80]).unique()

[(20.0, 60.0], NaN, (0.0, 20.0], (60.0, 80.0]]
Categories (3, interval[int64]): [(0, 20] < (20, 60] < (60, 80]]

In [110]:
df.Age.describe()

count    714.000000
mean      29.699118
std       14.526497
min        0.420000
25%       20.125000
50%       28.000000
75%       38.000000
max       80.000000
Name: Age, dtype: float64

## `.qcut()`

É utilizado quando queremos discretizar nossos dados em quantis. Ao informar quantidade de grupos o `pd.qcut()` escolhe os bins tal que tenhamos a mesma quantidade de valores em cada grupo.

`cut()` x `qcut()`  
- o comando `cut()` cria caixas equidistantes, mas a frequência das amostras é desigual em cada caixa  
- o comando `.qcut()` cria caixas de tamanhos desiguais, mas a frequência das amostras é igual em cada caixa

Parâmetros  
`pandas.qcut(x, q, labels=None, retbins=False, precision=3, duplicates='raise')`

# `Aula 4`

In [69]:
import pandas as pd
import numpy as np

In [70]:
#ler o df  do arquivo titanic.csv
tt = pd.read_csv('./data/titanic.csv', delimiter=',')
tt.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male 22.00      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00      1   
2                             Heikkinen, Miss. Laina  female 26.00      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00      1   
4                           Allen, Mr. William Henry    male 35.00      0   

   Parch            Ticket  Fare Cabin Embarked  
0      0         A/5 21171  7.25   NaN        S  
1      0          PC 17599 71.28   C85        C  
2      0  STON/O2. 3101282  7.92   NaN        S  
3      0            113803 53.10  C123        S  
4      0            373450  8.05   NaN        S

In [71]:
tt.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [1]:
# se quisermos calcular a média de fare por classe
for pclass in tt.Pclass.sort_values().unique():
    print(tt[tt['Pclass']==pclass].Fare.mean())

NameError: name 'tt' is not defined

e se quisermos calcular a classe por Pclass e Sex?

## `Groupby`

Assim como no SQL, no pandas também temos um método com o qual podemos agregar os dados. `o groupby` primeiro separa os dados em grupos definidos dentro do método, aplica um tipo de operação usando agregação, transformação, filtragem ou até uma função própria e, por fim junta os resultados encontrados.

<img src="groupby.png"  style="width: 700px" >

Exemplo de aplicação da função de agregação `mean`

Utilizar o Groupbyé o mesmo que fazer a sequência:
  
1. Dividir os dados em grupos utilizando um critério
2. Aplicar uma função em cada um dos grupos separadamente
3. combinar o resultado em uma estrutura de dados

## Funções de agregação
  
com essas funções podemos aplicar operações estatísticas nos nossos dados. Exemplos:
`mean`, `std`, `max`, `min`, `count`, `sum`, `var`.  
Quando queremos aplicar apenas uma dessas operações podemos chamá-lasdiretamente após o `groupby`:

In [73]:
tt.head()
tt.groupby(['Pclass']).Fare.mean()
#group by 'Pclass' and make mean of Fare

Pclass
1   84.15
2   20.66
3   13.68
Name: Fare, dtype: float64

In [74]:
tt.groupby(['Pclass'])['Fare'].mean()

Pclass
1   84.15
2   20.66
3   13.68
Name: Fare, dtype: float64

In [75]:
tt.groupby(['Pclass'])[['Fare']].mean()
# com dois coolchetes retorna um df

Fare
Pclass      
1      84.15
2      20.66
3      13.68

Opções de formatação

In [76]:
pd.options.display.float_format='{:,.2f}'.format

In [77]:
# Agrupa por Pclass e Sex e calcula a média de cada grupo
tt.groupby(['Pclass', 'Sex']).mean()

PassengerId  Survived   Age  SibSp  Parch   Fare
Pclass Sex                                                     
1      female       469.21      0.97 34.61   0.55   0.46 106.13
       male         455.73      0.37 41.28   0.31   0.28  67.23
2      female       443.11      0.92 28.72   0.49   0.61  21.97
       male         447.96      0.16 30.74   0.34   0.22  19.74
3      female       399.73      0.50 21.75   0.90   0.80  16.12
       male         455.52      0.14 26.51   0.50   0.22  12.66

Para calcular a média de uma só coluna

In [78]:
tt.groupby(['Pclass', 'Sex'])[['Age']].mean()
# tt.groupby(['Pclass', 'Sex']).mean()[['Age']]

Age
Pclass Sex         
1      female 34.61
       male   41.28
2      female 28.72
       male   30.74
3      female 21.75
       male   26.51

In [79]:
tt.groupby(['Pclass', 'Sex'])[['Age', 'Fare']].mean().index

MultiIndex([(1, 'female'),
            (1,   'male'),
            (2, 'female'),
            (2,   'male'),
            (3, 'female'),
            (3,   'male')],
           names=['Pclass', 'Sex'])

Quando queremos mais de uma operação chamamos o método `.agg()`

In [80]:
tt.groupby(['Pclass']).agg(['mean', 'max', 'min']).columns

MultiIndex([('PassengerId', 'mean'),
            ('PassengerId',  'max'),
            ('PassengerId',  'min'),
            (   'Survived', 'mean'),
            (   'Survived',  'max'),
            (   'Survived',  'min'),
            (        'Age', 'mean'),
            (        'Age',  'max'),
            (        'Age',  'min'),
            (      'SibSp', 'mean'),
            (      'SibSp',  'max'),
            (      'SibSp',  'min'),
            (      'Parch', 'mean'),
            (      'Parch',  'max'),
            (      'Parch',  'min'),
            (       'Fare', 'mean'),
            (       'Fare',  'max'),
            (       'Fare',  'min')],
           )

In [81]:
tt.groupby(['Pclass', 'Sex']).agg(['mean', 'max', 'min'])

PassengerId          Survived           Age            SibSp  \
                     mean  max min     mean max min  mean   max  min  mean   
Pclass Sex                                                                   
1      female      469.21  888   2     0.97   1   0 34.61 63.00 2.00  0.55   
       male        455.73  890   7     0.37   1   0 41.28 80.00 0.92  0.31   
2      female      443.11  881  10     0.92   1   0 28.72 57.00 2.00  0.49   
       male        447.96  887  18     0.16   1   0 30.74 70.00 0.67  0.34   
3      female      399.73  889   3     0.50   1   0 21.75 63.00 0.75  0.90   
       male        455.52  891   1     0.14   1   0 26.51 74.00 0.42  0.50   

                      Parch           Fare               
              max min  mean max min   mean    max   min  
Pclass Sex                                               
1      female   3   0  0.46   2   0 106.13 512.33 25.93  
       male     3   0  0.28   4   0  67.23 512.33  0.00  
2      female   3   0  0.61   3   0  21.97  65.00 10.50  
       male     2   0  0.22   2   0  19.74  73.50  0.00  
3      female   8   0  0.80   6   0  16.12  69.55  6.75  
       male     8   0  0.22   5   0  12.66  69.55  0.00

Para operações distintas em colunas distintas passamos um dicionario com o nome da coluna como chave e a operação como valor

In [82]:
tt.groupby(['Pclass']).agg(
    {'Embarked': pd.Series.mode, 'Fare':'mean'})

Embarked  Fare
Pclass               
1             S 84.15
2             S 20.66
3             S 13.68

A coluna utilizada no `groupby` virou um index do nosso df. Para convertê-la em coluna novamente temos duas formas:  
  
1. chamar o parâmetro `as_index=False` dentro do `groupby`
2. aplicar `.reset_index()` ao final da sentença

In [83]:
# método 1
tt.groupby(['Pclass'], as_index=False).agg({'Embarked': pd.Series.mode, 'Fare': 'mean'})

Pclass Embarked  Fare
0       1        S 84.15
1       2        S 20.66
2       3        S 13.68

In [84]:
# Método 2
tt.groupby(['Pclass']).agg({'Embarked':pd.Series.mode, 'Fare':'mean'}).reset_index()

Pclass Embarked  Fare
0       1        S 84.15
1       2        S 20.66
2       3        S 13.68

________________
________________
**Exercício**: Existe diferença de sobrevivência por portão de embarque? e difereça no preço do ticket? Pq acha que existe essa diferença?

In [85]:
tt.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [86]:
tt.groupby(['Embarked', 'Pclass'])[['Survived']].mean()

Survived
Embarked Pclass          
C        1           0.69
         2           0.53
         3           0.38
Q        1           0.50
         2           0.67
         3           0.38
S        1           0.58
         2           0.46
         3           0.19

In [87]:
tt.groupby(['Embarked', 'Pclass'])[['Fare']].mean()

Fare
Embarked Pclass       
C        1      104.72
         2       25.36
         3       11.21
Q        1       90.00
         2       12.35
         3       11.18
S        1       70.36
         2       20.33
         3       14.64

In [88]:
tt.corr()

PassengerId  Survived  Pclass   Age  SibSp  Parch  Fare
PassengerId         1.00     -0.01   -0.04  0.04  -0.06  -0.00  0.01
Survived           -0.01      1.00   -0.34 -0.08  -0.04   0.08  0.26
Pclass             -0.04     -0.34    1.00 -0.37   0.08   0.02 -0.55
Age                 0.04     -0.08   -0.37  1.00  -0.31  -0.19  0.10
SibSp              -0.06     -0.04    0.08 -0.31   1.00   0.41  0.16
Parch              -0.00      0.08    0.02 -0.19   0.41   1.00  0.22
Fare                0.01      0.26   -0.55  0.10   0.16   0.22  1.00

_______
_______

E se quisermos  uma coluna nova que contenham o valor médio do Fare por Pclass?

## Criando coluna com dado agregado
  
queremos que todas as pessoas da primeira classe tenham valor 84.15 nessa nova coluna, todas da segunda classe tenham valor 20.66 e da terceira classe 13.67.  
Podemos tentar:

In [89]:
tt.groupby('Pclass')[['Fare']].mean()

Fare
Pclass      
1      84.15
2      20.66
3      13.68

In [90]:
tt['Fare_mean'] = tt.groupby('Pclass')['Fare'].mean()

tt.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male 22.00      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00      1   
2                             Heikkinen, Miss. Laina  female 26.00      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00      1   
4                           Allen, Mr. William Henry    male 35.00      0   

   Parch            Ticket  Fare Cabin Embarked  Fare_mean  
0      0         A/5 21171  7.25   NaN        S        NaN  
1      0          PC 17599 71.28   C85        C      84.15  
2      0  STON/O2. 3101282  7.92   NaN        S      20.66  
3      0            113803 53.10  C123        S      13.68  
4      0            373450  8.05   NaN        S        NaN

``Xiii... deu ruim...``

## Transformação de dados  
  
Ao aplicarmos o método `.transform()` temos como retorno um objeto com o mesmo index do df de origem contendo a transformação realizada para cada uma das linhas. Dessa forma podemos utilizar esse método e apenas criar uma coluna mova para o nosso df.  
**Ele será muito útil na criação de novas features para os modelos.**

In [91]:
tt.groupby('Pclass')[['Fare']].mean()

Fare
Pclass      
1      84.15
2      20.66
3      13.68

In [92]:
tt.groupby('Pclass')[['Fare']].transform('mean')

Fare
0   13.68
1   84.15
2   13.68
3   84.15
4   13.68
..    ...
886 20.66
887 84.15
888 13.68
889 84.15
890 13.68

[891 rows x 1 columns]

In [93]:
tt['Fare_mean'] = tt.groupby('Pclass')['Fare'].transform('mean')
tt.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male 22.00      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00      1   
2                             Heikkinen, Miss. Laina  female 26.00      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00      1   
4                           Allen, Mr. William Henry    male 35.00      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male 54.00      0   
7                     Palsson, Master. Gosta Leonard    male  2.00      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female 27.00      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female 14.00      1   

   Parch            Ticket  Fare Cabin Embarked  Fare_mean  
0      0         A/5 21171  7.25   NaN        S      13.68  
1      0          PC 17599 71.28   C85        C      84.15  
2      0  STON/O2. 3101282  7.92   NaN        S      13.68  
3      0            113803 53.10  C123        S      84.15  
4      0            373450  8.05   NaN        S      13.68  
5      0            330877  8.46   NaN        Q      13.68  
6      0             17463 51.86   E46        S      84.15  
7      1            349909 21.07   NaN        S      13.68  
8      2            347742 11.13   NaN        S      13.68  
9      0            237736 30.07   NaN        C      20.66

podemos aplicar tanto as operações mencionadas na agregação quanto a uma função `lambda`

o que é uma função lambda?
I have no idea.

In [94]:
função_lambda = lambda x: 2*x
função_lambda(4)

8

In [97]:
tt['variação_max_min'] = tt.groupby('Pclass')[['Fare']].transform(lambda x: x.max() - x.min())
tt.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male 22.00      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00      1   
2                             Heikkinen, Miss. Laina  female 26.00      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00      1   
4                           Allen, Mr. William Henry    male 35.00      0   

   Parch            Ticket  Fare Cabin Embarked  Fare_mean  variação_max_min  
0      0         A/5 21171  7.25   NaN        S      13.68             69.55  
1      0          PC 17599 71.28   C85        C      84.15            512.33  
2      0  STON/O2. 3101282  7.92   NaN        S      13.68             69.55  
3      0            113803 53.10  C123        S      84.15            512.33  
4      0            373450  8.05   NaN        S      13.68             69.55

In [98]:
# conferindo a quantidade de nulos por coluna
tt.isna().sum()

PassengerId           0
Survived              0
Pclass                0
Name                  0
Sex                   0
Age                 177
SibSp                 0
Parch                 0
Ticket                0
Fare                  0
Cabin               687
Embarked              2
Fare_mean             0
variação_max_min      0
dtype: int64

In [99]:
# preencher nulos
tt[['Age']].mean()

Age   29.70
dtype: float64

In [101]:
tt.Age.fillna(tt.Age.mean())

0     22.00
1     38.00
2     26.00
3     35.00
4     35.00
       ... 
886   27.00
887   19.00
888   29.70
889   26.00
890   32.00
Name: Age, Length: 891, dtype: float64

In [102]:
tt['Age_sem_nulo'] = tt.groupby(['Pclass'])[['Age']].transform(lambda x: x.fillna(x.mean()))

In [103]:
tt.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male 22.00      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00      1   
2                             Heikkinen, Miss. Laina  female 26.00      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00      1   
4                           Allen, Mr. William Henry    male 35.00      0   

   Parch            Ticket  Fare Cabin Embarked  Fare_mean  variação_max_min  \
0      0         A/5 21171  7.25   NaN        S      13.68             69.55   
1      0          PC 17599 71.28   C85        C      84.15            512.33   
2      0  STON/O2. 3101282  7.92   NaN        S      13.68             69.55   
3      0            113803 53.10  C123        S      84.15            512.33   
4      0            373450  8.05   NaN        S      13.68             69.55   

   Age_sem_nulo  
0         22.00  
1         38.00  
2         26.00  
3         35.00  
4         35.00

In [105]:
tt.isna().sum()

PassengerId           0
Survived              0
Pclass                0
Name                  0
Sex                   0
Age                 177
SibSp                 0
Parch                 0
Ticket                0
Fare                  0
Cabin               687
Embarked              2
Fare_mean             0
variação_max_min      0
Age_sem_nulo          0
dtype: int64

In [106]:
# conferindo o preenchimento de nulos
# idade média por Pclass
tt.groupby(['Pclass'])[['Age']].mean()

Age
Pclass      
1      38.23
2      29.88
3      25.14

In [108]:
# selecionando a parte do df que tem idade nula
tt[tt.Age.isna()].head(10)

PassengerId  Survived  Pclass  \
5             6         0       3   
17           18         1       2   
19           20         1       3   
26           27         0       3   
28           29         1       3   
29           30         0       3   
31           32         1       1   
32           33         1       3   
36           37         1       3   
42           43         0       3   

                                              Name     Sex  Age  SibSp  Parch  \
5                                 Moran, Mr. James    male  NaN      0      0   
17                    Williams, Mr. Charles Eugene    male  NaN      0      0   
19                         Masselmani, Mrs. Fatima  female  NaN      0      0   
26                         Emir, Mr. Farred Chehab    male  NaN      0      0   
28                   O'Dwyer, Miss. Ellen "Nellie"  female  NaN      0      0   
29                             Todoroff, Mr. Lalio    male  NaN      0      0   
31  Spencer, Mrs. William Augustus (Marie Eugenie)  female  NaN      1      0   
32                        Glynn, Miss. Mary Agatha  female  NaN      0      0   
36                                Mamee, Mr. Hanna    male  NaN      0      0   
42                             Kraeff, Mr. Theodor    male  NaN      0      0   

      Ticket   Fare Cabin Embarked  Fare_mean  variação_max_min  Age_sem_nulo  
5     330877   8.46   NaN        Q      13.68             69.55         25.14  
17    244373  13.00   NaN        S      20.66             73.50         29.88  
19      2649   7.22   NaN        C      13.68             69.55         25.14  
26      2631   7.22   NaN        C      13.68             69.55         25.14  
28    330959   7.88   NaN        Q      13.68             69.55         25.14  
29    349216   7.90   NaN        S      13.68             69.55         25.14  
31  PC 17569 146.52   B78        C      84.15            512.33         38.23  
32    335677   7.75   NaN        Q      13.68             69.55         25.14  
36      2677   7.23   NaN        C      13.68             69.55         25.14  
42    349253   7.90   NaN        C      13.68             69.55         25.14

_________________________
_________________________
**Exercício:** Crie uma coluna com a média de Fare e outra com a média de idade para cada classe da coluna Survived. Você consegue fazer isso de uma única vez?

In [142]:
tt.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male 22.00      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00      1   
2                             Heikkinen, Miss. Laina  female 26.00      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00      1   
4                           Allen, Mr. William Henry    male 35.00      0   

   Parch            Ticket  Fare Cabin Embarked  Fare_mean  variação_max_min  \
0      0         A/5 21171  7.25   NaN        S      22.12             69.55   
1      0          PC 17599 71.28   C85        C      48.40            512.33   
2      0  STON/O2. 3101282  7.92   NaN        S      48.40             69.55   
3      0            113803 53.10  C123        S      48.40            512.33   
4      0            373450  8.05   NaN        S      22.12             69.55   

   Age_sem_nulo  
0         22.00  
1         38.00  
2         26.00  
3         35.00  
4         35.00